Exception trying to use a custom Layout

### Description

Exception trying to use a custom Layout
![image](https://user-images.githubusercontent.com/6755973/152794528-6fd27187-a604-4cc6-a5a9-68968c4632af.png)


### Steps to Reproduce

1. Download [LayoutIssue.zip](https://github.com/dotnet/maui/files/8015466/LayoutIssue.zip)
2. Launch the App in any platform.
 

### Version with bug

Preview 12 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS

### Affected platform versions

Windows SDK 10.0.17134.0, Android 11

### Did you find any workaround?

_No response_

### Relevant log output

```shell
[mono-rt] [ERROR] FATAL UNHANDLED EXCEPTION: System.Exception: Handler not found for view LayoutIssue.CustomLayout.
```

--- a/src/Compatibility/Core/src/AppHostBuilderExtensions.cs
+++ b/src/Compatibility/Core/src/AppHostBuilderExtensions.cs
@@ -133,6 +133,7 @@ namespace Microsoft.Maui.Controls.Hosting
 					handlers.TryAddCompatibilityRenderer(typeof(RefreshView), typeof(RefreshViewRenderer));
 					handlers.TryAddCompatibilityRenderer(typeof(NativeViewWrapper), typeof(NativeViewWrapperRenderer));
 
+					handlers.TryAddCompatibilityRenderer(typeof(Microsoft.Maui.Controls.Compatibility.Layout<View>), typeof(DefaultRenderer));
 					handlers.TryAddCompatibilityRenderer(typeof(Microsoft.Maui.Controls.Compatibility.RelativeLayout), typeof(DefaultRenderer));
 					handlers.TryAddCompatibilityRenderer(typeof(Microsoft.Maui.Controls.Compatibility.AbsoluteLayout), typeof(DefaultRenderer));
 

--- a/src/Controls/samples/Controls.Sample/Pages/Layouts/CustomLayoutPage.xaml
+++ b/src/Controls/samples/Controls.Sample/Pages/Layouts/CustomLayoutPage.xaml
@@ -1,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<views:BasePage 
+    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+    x:Class="Maui.Controls.Sample.Pages.CustomLayoutPage"
+    xmlns:layout="clr-namespace:Maui.Controls.Sample.Pages"
+    xmlns:views="clr-namespace:Maui.Controls.Sample.Pages.Base"
+    Title="Custom Layout">
+    <views:BasePage.Content>
+        <layout:DockLayout
+            LastChildFill="False">
+            <Button layout:DockLayout.Dock="Top" Text="Top" HeightRequest="50"/>
+            <Button layout:DockLayout.Dock="Bottom" Text="Bottom" HeightRequest="50"/>
+            <Button layout:DockLayout.Dock="Left" Text="Left" WidthRequest="60"/>
+            <Button layout:DockLayout.Dock="Left" Text="Left" WidthRequest="60"/>
+            <Button layout:DockLayout.Dock="Right" Text="Right" WidthRequest="80"/>
+            <Button layout:DockLayout.Dock="Right" Text="Right" WidthRequest="80"/>
+        </layout:DockLayout>
+    </views:BasePage.Content>
+</views:BasePage>

--- a/src/Controls/samples/Controls.Sample/Pages/Layouts/CustomLayoutPage.xaml.cs
+++ b/src/Controls/samples/Controls.Sample/Pages/Layouts/CustomLayoutPage.xaml.cs
@@ -1,0 +1,171 @@
+﻿using System;
+using Microsoft.Maui;
+using Microsoft.Maui.Controls;
+using Microsoft.Maui.Controls.Compatibility;
+using Microsoft.Maui.Graphics;
+
+namespace Maui.Controls.Sample.Pages
+{
+	public partial class CustomLayoutPage
+	{
+		public CustomLayoutPage()
+		{
+			InitializeComponent();
+		}
+	}
+
+	public enum Dock
+	{
+		Left,
+		Top,
+		Right,
+		Bottom
+	}
+
+	public class DockLayout : Layout<View>
+	{
+		public static readonly BindableProperty DockProperty =
+			BindableProperty.Create(nameof(Dock), typeof(Dock), typeof(DockLayout), Dock.Left,
+				BindingMode.TwoWay, null);
+
+		public Dock Dock
+		{
+			get { return (Dock)GetValue(DockProperty); }
+			set { SetValue(DockProperty, value); }
+		}
+
+		private Dock GetDock(BindableObject bindable)
+		{
+			return (Dock)bindable.GetValue(DockProperty);
+		}
+
+		public static readonly BindableProperty LastChildFillProperty =
+		   BindableProperty.Create(nameof(LastChildFill), typeof(bool), typeof(DockLayout), true,
+			   BindingMode.TwoWay, null);
+
+		/// <summary>
+		/// The default behavior is that the last child of the DockLayout takes up the rest of the space, 
+		/// but this can be disabled using the LastChildFill.
+		/// </summary>
+		public bool LastChildFill
+		{
+			get { return (bool)GetValue(LastChildFillProperty); }
+			set { SetValue(LastChildFillProperty, value); }
+		}
+
+		protected override void LayoutChildren(double x, double y, double width, double height)
+		{
+			SizeRequest sizeRequest = new SizeRequest();
+			int i = 0;
+
+			foreach (var child in Children)
+			{
+				if (child.IsVisible)
+				{
+					i++;
+
+					sizeRequest = child.Measure(width, height, MeasureFlags.IncludeMargins);
+
+					double childX = 0;
+					double childY = 0;
+					Size request = sizeRequest.Request;
+					double childWidth = Math.Min(width, request.Width);
+					double childHeight = Math.Min(height, request.Height);
+
+					bool lastItem = i == Children.Count;
+					if (lastItem & LastChildFill)
+					{
+						LayoutChildIntoBoundingRegion(child, new Rectangle(x, y, width, height));
+						return;
+					}
+
+					switch (GetDock(child))
+					{
+						case Dock.Left:
+							{
+								childX = x;
+								childY = y;
+								childHeight = height;
+								x += childWidth;
+								width -= childWidth;
+								break;
+							}
+						case Dock.Top:
+							{
+								childX = x;
+								childY = y;
+								childWidth = width;
+								y += childHeight;
+								height -= childHeight;
+								break;
+							}
+						case Dock.Right:
+							{
+								childX = x + width - childWidth;
+								childY = y;
+								childHeight = height;
+								width -= childWidth;
+								break;
+							}
+						case Dock.Bottom:
+							{
+								childX = x;
+								childY = y + height - childHeight;
+								childWidth = width;
+								height -= childHeight;
+								break;
+							}
+						default:
+							{
+								goto case Dock.Left;
+							}
+					}
+
+					LayoutChildIntoBoundingRegion(child, new Rectangle(childX, childY, childWidth, childHeight));
+				}
+			}
+		}
+
+		protected override SizeRequest OnMeasure(double widthConstraint, double heightConstraint)
+		{
+			double height = 0;
+			double width = 0;
+			double finalWidth = 0;
+			double finalHeight = 0;
+
+			foreach (var child in Children)
+			{
+				if (child.IsVisible)
+				{
+					SizeRequest sizeRequest = child.Measure(widthConstraint, heightConstraint, MeasureFlags.IncludeMargins);
+					Size request = sizeRequest.Request;
+
+					switch (GetDock(child))
+					{
+						case Dock.Left:
+						case Dock.Right:
+							{
+								width += request.Width;
+								finalWidth = Math.Max(finalWidth, width);
+								finalHeight = Math.Max(finalHeight, height + request.Height);
+								break;
+							}
+						case Dock.Top:
+						case Dock.Bottom:
+							{
+								height += request.Height;
+								finalWidth = Math.Max(finalWidth, width + request.Width);
+								finalHeight = Math.Max(finalHeight, height);
+								break;
+							}
+						default:
+							{
+								goto case Dock.Right;
+							}
+					}
+				}
+			}
+			return new SizeRequest(new Size(finalWidth, finalHeight));
+		}
+	}
+}

--- a/src/Controls/samples/Controls.Sample/ViewModels/LayoutsViewModel.cs
+++ b/src/Controls/samples/Controls.Sample/ViewModels/LayoutsViewModel.cs
@@ -40,13 +40,16 @@ namespace Maui.Controls.Sample.ViewModels
 				"A VerticalStackLayout organizes child views in a one-dimensional vertical stack."),
 
 			new SectionModel(typeof(LayoutUpdatesPage), "Layout Updates",
-				"Demonstrations of updating layouts"),
+				"Demonstrations of updating layouts."),
 
 			new SectionModel(typeof(ZIndexPage), "Z-Index",
-				"Demonstrations of the ZIndex property"),
+				"Demonstrations of the ZIndex property."),
 
 			new SectionModel(typeof(ClippingPage), "Clipping",
-				"Demonstrations of layout clipping"),
+				"Demonstrations of layout clipping."),
+			
+			new SectionModel(typeof(CustomLayoutPage), "Custom Layout",
+				"Demonstrations of custom layout."),
 		};
 	}
 }

